### PR TITLE
cpu/sam0_common/spi: demux SPI output pins before entering deep sleep

### PR DESCRIPTION
**Pull request by jue89**
### Contribution description

The output pins of `periph_spi` becomes HIGH-Z when entering a deep sleep power mode. 

If no external pull resistors are present, this will lead to undefined voltage levels and will create unwanted current draw by connected SPI slaves.
This can be observed with the `samr30` CPU as it has a sub-GHz transceiver inside its package that disables its pull resistor once it has been initialized.

### Testing procedure

Grab a `samr30-xpro` board and observe the current draw of the board with `examples/default`:

```
$ make BOARD=samr30-xpro flash term
2020-05-20 14:25:40,931 # main(): This is RIOT! (Version: 2020.04-devel-2669-g03322-fix/sam0_spi_deep-sleep)
2020-05-20 14:25:40,932 # Welcome to RIOT!
> ifconfig 4 set state SLEEP
2020-05-20 14:25:55,728 #  ifconfig 4 set state SLEEP
2020-05-20 14:25:55,732 # success: set state of interface 4 to SLEEP
> pm set 1
2020-05-20 14:26:03,830 #  pm set 1
2020-05-20 14:26:03,833 # CPU is entering power mode 1.
2020-05-20 14:26:03,836 # Now waiting for a wakeup event...
```

Without this patch, the board draws ~600uA. With this patch, it drops down to ~4uA.

If you want to test it with another `sam0_common`-based board, call `pm set 1` and try to pull the SPI pins with a resistor to VCC and GND. If this works, you verified these pins to be HIGH-Z.

### CODE REVIEW

File: cpu/sam0_common/periph/spi.c [hunk 1]
```diff
@@ -184,3 +184,25 @@ void spi_transfer_bytes(spi_t bus, spi_cs_t cs, bool cont,
...
+    }
+}
+
+void spi_pm_cb_leave(int deep)
+{
+    if (deep) {
+        for (size_t i = 0; i < SPI_NUMOF; i++) {
+            spi_t bus = SPI_DEV(i);
+            gpio_init_mux(spi_config[bus].mosi_pin, spi_config[bus].mosi_mux);
+            gpio_init_mux(spi_config[bus].clk_pin, spi_config[bus].clk_mux);
```
Comment (benpicco) [inline cpu/sam0_common/periph/spi.c:205]: The problem here is that not all SPI buses from the board configuration are necessarily used by the application, so those pins might be used for something else.

I don't see any register that we can reliably query here, so how about introducing a `static uint8_t spi_enabled;` bit field that gets set in `spi_init()`

    spi_enabled |= 1 << bus;

Then you can skip those entries that are not enabled in the loop here and in `spi_pm_cb_enter`.

Comment (jue89) [reply cpu/sam0_common/periph/spi.c:205]: Good idea. This bitmask may become useful, once `periph_spi_reconfigure` has been merged and `spi_deinit_pins()` gets called.

**Comment by bergzand** [comment]
Does it make sense to instead do this in the platform `spi_acquire` and `spi_release` functions? Not sure what the overhead is of adding it there, but the impact might be less than changing this every time the MCU sleeps

**Comment by benpicco** [comment]
> Not sure what the overhead is of adding it there

Luckily `tests/periph_spi` includes a benchmark :wink:

**Comment by bergzand** [comment]
> Luckily `tests/periph_spi` includes a benchmark wink

Unfortunately, `tests/periph_spi` doesn't measure the acquire/release cycle (could be added though :wink: )

**Comment by jue89** [comment]
> Does it make sense to instead do this in the platform `spi_acquire` and `spi_release` functions? Not sure what the overhead is of adding it there, but the impact might be less than changing this every time the MCU sleeps

Hmm, in this case, we should interact with `pm_layered` to prevent the MCU from falling into deep sleep?! I think the `stm32` are already have something like this implemented ...

**Comment by benpicco** [comment]
I think @bergzand was suggesting to demux after release, not sure if that messes with the slave though.

**Comment by jue89** [comment]
> I think @bergzand was suggesting to demux in `spi_release`, not sure if that messes with the slave though.

I'll try this solution as well ;)

My idea was to `pm_block(PM_STANDBY)` in `spi_acquire()` and `pm_unblock(PM_STANDBY)` in `spi_release()`. This ensures that the MCU doesn't fall in deep sleep within a SPI transaction.

**Comment by benpicco** [comment]
> My idea was to `pm_block(PM_STANDBY)` in `spi_acquire()` and `pm_unblock(PM_STANDBY)` in `spi_release()`. This ensures that the MCU doesn't fall in deep sleep within a SPI transaction.

That's a good idea too, will be useful when adding DMA.

File: cpu/sam0_common/periph/spi.c [hunk 2]
```diff
@@ -184,3 +197,29 @@ void spi_transfer_bytes(spi_t bus, spi_cs_t cs, bool cont,
...
+                gpio_disable_mux(spi_config[bus].mosi_pin);
+                gpio_disable_mux(spi_config[bus].clk_pin);
+            }
+        }
+    }
+}
+
+void spi_pm_cb_leave(int deep)
+{
+    if (deep) {
```
Comment (benpicco) [inline cpu/sam0_common/periph/spi.c:216]: Minor nit:

```suggestion
    if (!deep) {
        return;
    }
```

saves you unnececary nesting.

**Comment by jue89** [comment]
And the alternative #14107 also works :)

If we decide to go this way, I'll PR the interaction with `pm_layered`.

**Comment by jue89** [comment]
After some digging around, I came to the conclusion that this behavior is not related to the PM!
If the SPI peripheral is disabled (what happens during `spi_release()`), all output pins go HIGH-Z! It just becomes obvious during PM ... so I'll close this PR and fix this issue in #14107.

**Pull request closed without merging**